File: 프로그래머스/2/12978. 배달/배달.cpp
#include <vector>

#define MAX 500001

using namespace std;

int solution(int N, vector<vector<int> > roads, int K) {
    int answer = 0;
    int map[50][50];
    for(int j=0;j<N;j++){
        for(int k=0;k<N;k++){
            if(j!=k)
                map[j][k] = MAX;
            else
                map[j][k] = 0;
        }
    }
    for(auto road : roads){
        if(map[road[0]-1][road[1]-1] > road[2] || map[road[1]-1][road[0]-1] > road[2]){
            map[road[0]-1][road[1]-1] = road[2];
            map[road[1]-1][road[0]-1] = road[2];            
        }
    }
    // i -> 거쳐가는 노드 / j -> 출발 노드 / k -> 도착 노드
    for(int i=0;i<N;i++){
        for(int j=0;j<N;j++){
            for(int k=0;k<N;k++){
                if(map[j][k] > map[j][i] + map[i][k])
                    map[j][k] = map[j][i] + map[i][k];
            }
        }
    }    
    for(int i=0;i<N;i++){
        if(map[0][i] <= K)
            answer++;
    }
        
    return answer;
}

File: 프로그래머스/2/12978. 배달/배달.py
def solution(N, road, K):
    max_distance = 500000 * 20 + 1
    
    ## 초기화
    road_map = [[max_distance if i!=j else 0 for j in range(N)] for i in range(N)]
    
    ## 그래프 길이 저장
    for r in road:
        s, d, l = r
        road_map[s-1][d-1] = min(l, road_map[s-1][d-1])
        road_map[d-1][s-1] = min(l, road_map[d-1][s-1])
    
    # print(road_map)
    
    ## Floyed Washal
    ## i -> 거쳐가는 노드 / j -> 출발 노드 / k -> 도착 노드
    for i in range(N):
        for j in range(N):
            for k in range(N):
                if road_map[j][k] > road_map[j][i] + road_map[i][k]:
                    road_map[j][k] = min(road_map[j][k] , road_map[j][i] + road_map[i][k])
    
    # print(road_map)
    # print(road_map[0])
    return sum([r <= K for r in road_map[0]])